sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala

```diff
@@ -255,61 +254,41 @@ private[sql] trait SQLTestUtilsBase
    * Drops temporary view `viewNames` after calling `f`.
    */
   protected def withTempView(viewNames: String*)(f: => Unit): Unit = {
-    try f finally {
-      // If the test failed part way, we don't want to mask the failure by failing to remove
-      // temp views that never got created.
-      try viewNames.foreach(spark.catalog.dropTempView) catch {
-        case _: NoSuchTableException =>
-      }
-    }
+    Utils.tryWithSafeFinally(f)(viewNames.foreach(spark.catalog.dropTempView))
```

Comment:
This does change behavior slightly; NoSuchTableException is no longer suppressed. You may have to keep that part here and elsewhere.

Reply:
Hi @srowen , 

Yes. the code change the behaviour slightly. The original code would completely ignore the `NoSuchTableException`. The new one (via `Utils.tryWithSafeFinally`) would add the `NoSuchTableException` as a suppressed exception to the original exception or simply rethrow the `NoSuchTableException` if we did not hit any exception in the try block. 

I feel like the new behaviour is more desirable. Otherwise, it will be very easy for a  developer to miss any typo on view names as the test would not fail even if the provided view name is never correct.

Somehow, the original code suppresses the `NoSuchTableException` outside the foreach loop. Therefore, the loop would be terminated immediately when hitting the `NoSuchTableException`. If it is not the desirable behaviour, I guess the original code should be something like

```
    viewNames.foreach { viewName =>
      try spark.catalog.dropTempView(viewName) catch {
        // If the test failed part way, we don't want to mask the failure by failing to remove
        // temp views that never got created.
        case _: NoSuchTableException =>
      }
    }
```

What about I change the code into something like this? 

```
    Utils.tryWithSafeFinally(f) {
      var firstException: Throwable = null
      viewNames.foreach { viewName =>
        try spark.catalog.dropTempView(viewName) catch {
          case t: Throwable =>
            if (firstException == null) {
              firstException = t
            } else {
              firstException.addSuppressed(t)
            }
        }
      }
      if (firstException != null) {
        throw firstException
      }
    }
```
However, the code looks over complicated and very procedure... I don't like this either.

Reply:
A slightly modified version. Less code and no more null. 

```
Utils.tryWithSafeFinally(f) {
      val suppressed = viewNames.map(viewName => Try(spark.catalog.dropTempView(viewName))).map {
        case Failure(t) => t
      }
      if (suppressed.nonEmpty) {
        suppressed.tail.foreach(suppressed.head.addSuppressed)
        throw suppressed.head
      }
    }
```

Reply:
It looks like this code purposely doesn't throw NoSuchTableException in any case. It doesn't matter much if it's just added as a suppressed exception, but in the case the main block doesn't throw, NoSuchTableException is thrown. I wouldn't change the semantics here without being sure; it's not clear to me it indicates user error.

Why not just leave the finally block as is? it's simpler.